Assignment5/Q1.py

```diff
@@ -0,0 +1,150 @@
+
+        Returns:
+            graph: a dictionary of sets, corresponds to graph data structure, where characters from dictionary are
+                   'vertices' and 'edges' indicate characters' partial ordering
+            state: a dictionary of strings, where each string is either "white", "grey" or "black". It indicates the
+                   state that the "vertex" (character) is in while topological search is performed. In this method
+                   only initialize state dictionary to all "white".
+    """
+    graph = {}
+    state = {}
+
+    # Instantiate the graph data structure and set state to "white" for all characters/,
+    for word in dictionary:
+        for char in word:
+            if char not in graph:
+                graph[char] = set()
+                state[char] = "white"
+
+    # Set the orderings of characters based on the order of words in given dictionary.
+    # (NOTE: it is enough to compare only adjacent words).
+    for i in range(len(dictionary) - 1):
+        adjacent_word1 = dictionary[i]
+        adjacent_word2 = dictionary[i + 1]
+        index = get_index_of_first_differing_char(adjacent_word1, adjacent_word2)
+        if index != -1:
+            graph[adjacent_word1[index]].add(adjacent_word2[index])
+
+    return graph, state
+
+
+def find_total_order_from_char(partial_order, state, char, total_order):
+    """This method is the recursive part of topological sort (Similar to DFS from vertex).
+
+        Args:
+            partial_order: a dictionary of sets, graph data structure corresponding to characters partial order
+            state: a dictionary of strings, where strings indicate state of the character ("vertex")
+                   white - the character ("vertex") has not been visited by the search yet
+                   grey - the character ("vertex") is being visited, recursive call on it is not finished yet
+                   black - the character ("vertex") has already been visited
+            char: a character, from which the search is performed
+            total_order: a singly linked list of characters, the total order (lexicographical order) "accumulated
+                         so far" the new character is always added at the front of the list.
+    """
+    # The character is in search now.
+    state[char] = "grey"
+
+    # Check all characters that are "next" in the partial ordering.
+    for following_char in partial_order[char]:
+        # Perform a search from this character if it has not been visited yet.
+        if state[following_char] == "white":
+            find_total_order_from_char(partial_order, state, following_char, total_order)
+        # If any searched characters are found here, we found a cycle in the order, i.e. it is not a correct ordering.
+        elif state[following_char] == "grey":
+            raise ValueError("The given dictionary is not sorted in lexicographical order!")
+        break
```

Comment:
Absolutely not!! I think I wanted to break if the error occurs but I must have forgotten that error exits the function immediately. I removed the break, it was my mistake.